### PR TITLE
executor: Check for column mismatch

### LINKED ISSUE #41410: LOAD DATA LOCAL INFILE results in "commit forced to quit"
### 1. Minimal reproduce step (Required)

```
sql> \! cat /tmp/t/test.t1.0001.csv
"id","c1"
"1",0x010203
"2",0x020304
"3",0x030405

sql> CREATE TABLE t1 (id int primary key, c1 VARBINARY(255));
Query OK, 0 rows affected (0.0162 sec)

sql> LOAD DATA LOCAL INFILE '/tmp/t/test.t1.0001.csv' INTO TABLE t1 FIELDS TERMINATED BY ',' ENCLOSED BY '"' IGNORE 1 LINES (c1, c2);
ERROR: 1105 (HY000): commit forced to quit
```

### 2. What did you expect to see? (Required)

Error about column name mismatch

### 3. What did you see instead (Required)

```
ERROR: 1105 (HY000): commit forced to quit
```

```
[2023/02/14 14:11:07.673 +01:00] [ERROR] [load_data.go:215] ["process routine panicked"] [conn=2199023255957] [r="\"invalid memory address or nil pointer dereference\""] [stack="github.com/pingcap/tidb/executor.processStream.func1\n\t/home/dvaneeden/dev/pingcap/tidb/executor/load_data.go:217\nruntime.gopanic\n\t/usr/lib/golang/src/runtime/panic.go:884\nruntime.panicmem\n\t/usr/lib/golang/src/runtime/panic.go:260\nruntime.sigpanic\n\t/usr/lib/golang/src/runtime/signal_unix.go:835\ngithub.com/pingcap/tidb/executor.(*LoadDataInfo).colsToRow\n\t/home/dvaneeden/dev/pingcap/tidb/executor/load_data.go:813\ngithub.com/pingcap/tidb/executor.(*LoadDataInfo).InsertData\n\t/home/dvaneeden/dev/pingcap/tidb/executor/load_data.go:732\ngithub.com/pingcap/tidb/executor.insertDataWithCommit\n\t/home/dvaneeden/dev/pingcap/tidb/executor/load_data.go:274\ngithub.com/pingcap/tidb/executor.processStream\n\t/home/dvaneeden/dev/pingcap/tidb/executor/load_data.go:251"]
[2023/02/14 14:11:07.673 +01:00] [ERROR] [load_data.go:522] ["commit forced to quit, possible preparation failed"] [conn=2199023255957]
[2023/02/14 14:11:07.673 +01:00] [INFO] [conn.go:1632] ["not drained yet, try reading left data from client connection"] [conn=2199023255957]
[2023/02/14 14:11:07.673 +01:00] [INFO] [conn.go:1648] ["draining finished for error"] [conn=2199023255957] [error="commit forced to quit"]
[2023/02/14 14:11:07.673 +01:00] [INFO] [conn.go:1158] ["command dispatched failed"] [conn=2199023255957] [connInfo="id:2199023255957, addr:[::1]:38016 status:10, collation:utf8mb4_0900_ai_ci, user:root"] [command=Query] [status="inTxn:0, autocommit:1"] [sql="LOAD DATA LOCAL INFILE '/tmp/t/test.t1.0001.csv' INTO TABLE t1 FIELDS TERMINATED BY ',' ENCLOSED BY '\"' IGNORE 1 LINES (c1, c2)"] [txn_mode=PESSIMISTIC] [timestamp=439453028888608768] [err="commit forced to quit\ngithub.com/pingcap/tidb/executor.(*LoadDataInfo).CommitWork\n\t/home/dvaneeden/dev/pingcap/tidb/executor/load_data.go:521\ngithub.com/pingcap/tidb/executor.(*LoadDataInfo).Load\n\t/home/dvaneeden/dev/pingcap/tidb/executor/load_data.go:202\ngithub.com/pingcap/tidb/server.(*clientConn).handleLoadData\n\t/home/dvaneeden/dev/pingcap/tidb/server/conn.go:1629\ngithub.com/pingcap/tidb/server.(*clientConn).handleQuerySpecial\n\t/home/dvaneeden/dev/pingcap/tidb/server/conn.go:2048\ngithub.com/pingcap/tidb/server.(*clientConn).handleStmt\n\t/home/dvaneeden/dev/pingcap/tidb/server/conn.go:2027\ngithub.com/pingcap/tidb/server.(*clientConn).handleQuery\n\t/home/dvaneeden/dev/pingcap/tidb/server/conn.go:1849\ngithub.com/pingcap/tidb/server.(*clientConn).dispatch\n\t/home/dvaneeden/dev/pingcap/tidb/server/conn.go:1380\ngithub.com/pingcap/tidb/server.(*clientConn).Run\n\t/home/dvaneeden/dev/pingcap/tidb/server/conn.go:1129\ngithub.com/pingcap/tidb/server.(*Server).onConn\n\t/home/dvaneeden/dev/pingcap/tidb/server/server.go:625\nruntime.goexit\n\t/usr/lib/golang/src/runtime/asm_amd64.s:1594"]
```

```
github.com/pingcap/tidb/executor.processStream.func1
	/home/dvaneeden/dev/pingcap/tidb/executor/load_data.go:217
runtime.gopanic
	/usr/lib/golang/src/runtime/panic.go:884
runtime.panicmem
	/usr/lib/golang/src/runtime/panic.go:260
runtime.sigpanic
	/usr/lib/golang/src/runtime/signal_unix.go:835
github.com/pingcap/tidb/executor.(*LoadDataInfo).colsToRow
	/home/dvaneeden/dev/pingcap/tidb/executor/load_data.go:813
github.com/pingcap/tidb/executor.(*LoadDataInfo).InsertData
	/home/dvaneeden/dev/pingcap/tidb/executor/load_data.go:732
github.com/pingcap/tidb/executor.insertDataWithCommit
	/home/dvaneeden/dev/pingcap/tidb/executor/load_data.go:274
github.com/pingcap/tidb/executor.processStream
	/home/dvaneeden/dev/pingcap/tidb/executor/load_data.go:251
```

```
commit forced to quit
github.com/pingcap/tidb/executor.(*LoadDataInfo).CommitWork
	/home/dvaneeden/dev/pingcap/tidb/executor/load_data.go:521
github.com/pingcap/tidb/executor.(*LoadDataInfo).Load
	/home/dvaneeden/dev/pingcap/tidb/executor/load_data.go:202
github.com/pingcap/tidb/server.(*clientConn).handleLoadData
	/home/dvaneeden/dev/pingcap/tidb/server/conn.go:1629
github.com/pingcap/tidb/server.(*clientConn).handleQuerySpecial
	/home/dvaneeden/dev/pingcap/tidb/server/conn.go:2048
github.com/pingcap/tidb/server.(*clientConn).handleStmt
	/home/dvaneeden/dev/pingcap/tidb/server/conn.go:2027
github.com/pingcap/tidb/server.(*clientConn).handleQuery
	/home/dvaneeden/dev/pingcap/tidb/server/conn.go:1849
github.com/pingcap/tidb/server.(*clientConn).dispatch
	/home/dvaneeden/dev/pingcap/tidb/server/conn.go:1380
github.com/pingcap/tidb/server.(*clientConn).Run
	/home/dvaneeden/dev/pingcap/tidb/server/conn.go:1129
github.com/pingcap/tidb/server.(*Server).onConn
	/home/dvaneeden/dev/pingcap/tidb/server/server.go:625
runtime.goexit
	/usr/lib/golang/src/runtime/asm_amd64.s:1594
```

### 4. What is your TiDB version? (Required)

```
sql> SELECT tidb_version()\G
*************************** 1. row ***************************
tidb_version(): Release Version: v6.6.0-alpha-500-ga78ff88c3-dirty
Edition: Community
Git Commit Hash: a78ff88c3d6916662963c6f8ab4ac693b714dd51
Git Branch: master
UTC Build Time: 2023-02-14 13:03:23
GoVersion: go1.19.5
Race Enabled: false
TiKV Min Version: 6.2.0-alpha
Check Table Before Drop: false
Store: unistore
1 row in set (0.0011 sec)
```

### PR DESCRIPTION
**Pull request by dveeden**
### What problem does this PR solve?

Issue Number: close #41410 

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [x] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [x] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
LOAD DATA INFILE now handles the situation where the column names of the table don't match the named columns in the statement
```

### CODE REVIEW

File: executor/load_data.go [hunk 1]
```diff
@@ -330,13 +330,11 @@ func (e *LoadDataInfo) initLoadColumns(columnNames []string) error {
 		for _, v := range e.ColumnAssignments {
 			columnNames = append(columnNames, v.Column.Name.O)
 		}
+	}
 
-		cols, missingColName = table.FindCols(tableCols, columnNames, e.Table.Meta().PKIsHandle)
-		if missingColName != "" {
-			return errors.Errorf("LOAD DATA INTO %s: unknown column %s", e.Table.Meta().Name.O, missingColName)
```
Comment (lance6716) [inline executor/load_data.go:336]: seems this function `initLoadColumns` handles the LOAD DATA ... SET col=xxx part, and `initFieldMappings` handles LOAD DATA ... (col1, col2) ... part. Maybe we can move the check to `initFieldMappings`.

Comment (dveeden) [reply executor/load_data.go:336]: What benefit would that give us?

Comment (lance6716) [reply executor/load_data.go:336]: not important. I tested the behaviour of different places the unknown column appears in MySQL:

```
CREATE TABLE `t` (
  `c` int(11) DEFAULT NULL,
  `c2` int(11) DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=latin1
```

```
MySQL [test]> load data local infile '/tmp/t.csv' into table t fields terminated by ',' (@v1, xxx1) set c=xxx2, c2=@v1;
ERROR 1054 (42S22): Unknown column 'xxx1' in 'field list'
```

although `c` is the first column in table and it's assigned from `xxx2`, MySQL thinks the first unknown column is `xxx1`. I guess we should process `... (col1, col2, ...) ...` part before `... SET col=xxx` and separately. Not sure that's over-imitation.


And the error message should also be changed.

Comment (dveeden) [reply executor/load_data.go:336]: I fixed the message, it now looks like this:

Comment (dveeden) [reply executor/load_data.go:336]: I've created #41468 for another case with a different error message and error number.

**Review by lance6716 — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**